File: sdk/python/mcp_observability.py
import requests
import uuid
import datetime

class MCPObservability:
    def __init__(self, api_url: str):
        self.api_url = api_url.rstrip('/')

    def trace(self, task, context, model_output, metadata=None):
        trace_id = str(uuid.uuid4())
        payload = {
            'id': trace_id,
            'timestamp': datetime.datetime.utcnow().isoformat() + 'Z',
            'task': task,
            'context': context,
            'model_output': model_output,
            'metadata': metadata or {}
        }
        resp = requests.post(f'{self.api_url}/traces', json=payload)
        resp.raise_for_status()
        return resp.json()

# Example usage:
# obs = MCPObservability(api_url='http://localhost:3001')
# obs.trace('Summarize', {'userMessage': 'Summarize the last 5 emails'}, 'Here is the summary...', {'runtimeMs': 314, 'model': 'gpt-4', 'retries': 0})

File: sdk/typescript/mcp-observability.ts
import axios from 'axios';
import { v4 as uuidv4 } from 'uuid';

export interface TraceMetadata {
  [key: string]: any;
}

export async function trace({
  apiUrl,
  apiKey,
  task,
  context,
  modelOutput,
  metadata = {}
}: {
  apiUrl: string;
  apiKey: string;
  task: string;
  context: any;
  modelOutput: string;
  metadata?: TraceMetadata;
}) {
  if (!apiKey) {
    console.warn('No API key provided. Trace will not be sent.');
    return null;
  }

  const traceId = uuidv4();
  const payload = {
    id: traceId,
    timestamp: new Date().toISOString(),
    task,
    context,
    model_output: modelOutput,
    metadata
  };

  const headers = {
    'Content-Type': 'application/json',
    'X-API-Key': apiKey
  };

  try {
    const resp = await axios.post(`${apiUrl.replace(/\/$/, '')}/traces`, payload, { headers });
    return resp.data;
  } catch (error) {
    console.error('Failed to send trace:', error);
    return null;
  }
}

// Example usage:
// await trace({
//   apiUrl: 'http://localhost:3001',
//   apiKey: 'mcp_your_api_key_here', // Get this from your dashboard
//   task: 'Summarize',
//   context: { userMessage: 'Summarize the last 5 emails' },
//   modelOutput: 'Here is the summary...',
//   metadata: { runtimeMs: 314, model: 'gpt-4', retries: 0 }
// });
